java/connector-node/risingwave-sink-mock-flink/risingwave-sink-mock-flink-runtime/src/main/java/com/risingwave/connector/FlinkDynamicAdapterFactory.java

```diff
@@ -0,0 +1,132 @@
+ * limitations under the License.
+ */
+
+package com.risingwave.connector;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.risingwave.connector.api.FlinkDynamicAdapterConfig;
+import com.risingwave.connector.api.FlinkSinkValidateAndWriterFactory;
+import com.risingwave.connector.api.TableSchema;
+import com.risingwave.connector.api.sink.SinkFactory;
+import com.risingwave.connector.api.sink.SinkWriter;
+import com.risingwave.connector.api.sink.SinkWriterV1;
+import com.risingwave.connector.context.DynamicTableSinkContextImpl;
+import com.risingwave.connector.sinkwriter.SinkWriterImpl;
+import com.risingwave.connector.sinkwriter.SinkWriterV2Impl;
+import com.risingwave.proto.Catalog;
+import java.util.*;
+import org.apache.flink.api.connector.sink.Sink;
+import org.apache.flink.configuration.ConfigOption;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.table.api.Schema;
+import org.apache.flink.table.api.TableException;
+import org.apache.flink.table.catalog.*;
+import org.apache.flink.table.connector.sink.*;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.factories.DynamicTableSinkFactory;
+import org.apache.flink.table.factories.FactoryUtil;
+
+public class FlinkDynamicAdapterFactory implements SinkFactory {
+
+    FlinkSinkValidateAndWriterFactory flinkSinkValidateAndWriterFactory;
+
+    public FlinkDynamicAdapterFactory(
+            FlinkSinkValidateAndWriterFactory flinkSinkValidateAndWriterFactory) {
+        this.flinkSinkValidateAndWriterFactory = flinkSinkValidateAndWriterFactory;
+    }
+
+    // Create corresponding Sink according to different configurations.
+    public DynamicTableSink.SinkRuntimeProvider buildDynamicTableSinkProvider(
+            TableSchema tableSchema, FlinkDynamicAdapterConfig config) {
+        List<Column> flinkColumns = FlinkDynamicAdapterUtil.getFlinkColumnsFromSchema(tableSchema);
+
+        // Start with the default value, and add as needed later
+        ObjectIdentifier objectIdentifier =
+                ObjectIdentifier.of("catalog_name", "database_name", "obj_name");
+        ClassLoader contextClassLoader = Thread.currentThread().getContextClassLoader();
+        DynamicTableSinkFactory dynamicTableSinkFactory =
+                flinkSinkValidateAndWriterFactory.getDynamicTableSinkFactory();
+        Set<ConfigOption<?>> configOptions = new HashSet<>();
+        configOptions.addAll(dynamicTableSinkFactory.requiredOptions());
+        configOptions.addAll(dynamicTableSinkFactory.optionalOptions());
+        config.processOption(configOptions);
+
+        Schema.Builder schemaBuilder = Schema.newBuilder();
+        tableSchema
+                .getColumnDescs()
+                .forEach(
+                        (columnDesc) ->
+                                schemaBuilder.column(
+                                        columnDesc.getName(), columnDesc.getDataType().toString()));
```

Comment:
It's better to first convert the rwDataType to flink data type and then pass the flink data type to the schema builder.